docs/source/concepts/InstrumentAccessLayers.rst

```diff
@@ -31,216 +33,170 @@ ____________
+    throw runtime_error("There is no instrument in input workspace.")
 
-The ``DetectorInfo`` object is accessed by an index going from 0 to the number of detectors. A vector of detector IDs can be obtained from a call to ``detectorInfo.detectorIDs()``.
+  for (size_t i = 0; i < componentInfo.size(); ++i) {
+    componentInfo.solidAngle(i, observer);
+  }
 
-It is also possible to use the method ``detectorInfo.indexOf(detID)`` to get the index for a particular detector ID. However, this is a call to a lookup in an unordered map, so is an expensive calculation which should be avoided where possible.
+Access to Components and working with Detectors
+_______________________________________________
 
-Below is an example refactoring.
+Detector Indices are the same as the corresponding Component Indices. Note that there are no dynamic casts.
 
-**Before refactoring**
+**Combining DetectorInfo and ComponentInfo**
 
 .. code-block:: c++
 
-  auto instrument = ws->getInstrument();
-  if (!instrument)
-    throw runtime_error("There is no instrument in input workspace.")
+  #include "MantidGeometry/Instrument/ComponentInfo.h"
+  #include "MantidGeometry/Instrument/DetectorInfo.h"
 
-  size_t numdets = instrument->getNumberDetectors();
-  vector<detid_t> detIds = instrument->getDetectorIDs();
+  ...
 
-  for (size_t i = 0; i < numdets; ++i) {
-    IDetector_const_sptr tmpdetector = instrument->getDetector(detIds[i]);
-    if (tmpdetector->isMasked()) {
-      maskeddetids.push_back(tmpdetid);
+  const auto &componentInfo = ws->componentInfo();
+  const auto &detectorInfo = ws->componentInfo();
+  if (componentInfo.size() == 0)
+    throw runtime_error("There is no instrument in input workspace.")
+  
+  std::vector<double> solidAnglesForDetectors(detectorInfo.size(), -1.0);
+  for (size_t i = 0; i < componentInfo.size(); ++i) {
+    if(componentInfo.isDetector(i) && !detectorInfo.isMasked(i)) 
+     solidAnglesForDetectors[i] = componentInfo.solidAngle(i, observer);
     }
   }
 
-**After - looping over index**
+``ComponentInfo`` can give quick access to parent and sub-tree component and detector indices.
 
 .. code-block:: c++
 
-  #include "MantidAPI/Detector.h"
+  #include "MantidGeometry/Instrument/ComponentInfo.h"
+  #include "MantidGeometry/Instrument/DetectorInfo.h"
 
   ...
 
-  const auto &detectorInfo = ws->detectorInfo();
-  if (detectorInfo.size() == 0)
+  const auto &componentInfo = ws->componentInfo();
+  const auto &detectorInfo = ws->componentInfo();
+  if (componentInfo.size() == 0)
     throw runtime_error("There is no instrument in input workspace.")
-
-  vector<detid_t> detIds = detectorInfo.detectorIDs();
-
-  for (size_t i = 0; i < detectorInfo.size(); ++i) {
-    if (detectorInfo.isMasked(i)) {
-      maskedDetIds.push_back(detIds[i]);
+  
+  std::set<size_t> componentsHoldingOnlyDetectors;
+  for (size_t i = 0; i < componentInfo.size(); ++i) {
+    if(componentInfo.componentsInSubtree(i) == componentInfo.detectorsInSubtree(i)) {
```

Comment:
Doesn't `componentsInSubtree` include itself, so this example will simply give you a list of detectors, i.e., nothing useful?